### ISSUE TITLE
From which function can I get shape functions and their gradients at the position of nodes?

### ISSUE THREAD

**Issue by Vahid-Galavi** [KratosMultiphysics/Kratos#6646]
Hi,

I need to define integration points in an element at the position of nodes (Newton-Cotes integration). I checked the class geometry and it couldn't find Newton-Cotes integration. Is it available in Kratos?

Thanks for your help!

Vahid

**Comment by loumalouomega** [comment]
> Newton-Cotes

No, but they can be defined. For which geometries do you want them?

**Comment by Vahid-Galavi** [comment]
At the moment for line geometries. I can also define them inside the element but I think it would be nice to have them in the core.

**Comment by loumalouomega** [comment]
Well, integration are defined in geometries, for 2D line:

https://github.com/KratosMultiphysics/Kratos/blob/104534058c90576393d45e4dc347b8e4f410e897/kratos/geometries/line_2d_2.h#L1192

As you see, line already has the maximum number of integration schemes (10). Right now is quite rigid. It is defined in compilation time and maximum number is 10. @pooyan-dadvand, any suggestion?

**Comment by loumalouomega** [comment]
BTW, if necessary implementation can be found here: https://people.sc.fsu.edu/~jburkardt/f_src/line_nco_rule/line_nco_rule.html

**Comment by pooyan-dadvand** [comment]
I think is not a bad idea to add it as a new IntegrationMethod to the corresponding geometries. @loumalouomega, if I remember correctly you have added alternative methods to geometries before.

**Comment by loumalouomega** [comment]
> I think is not a bad idea to add it as a new IntegrationMethod to the corresponding geometries. @loumalouomega, if I remember correctly you have added alternative methods to geometries before.

Yes, in fact the extended methods were added by me. The thing is that is limited to 10 integration methods, 5 standard and 5 extended. The system is a little bit rigid, and more schemes can be added, but maybe instead of patching and adding new integration methods following this process, we can add a macro in order to generalize or something less "handcrafted"

**Comment by pooyan-dadvand** [comment]
If you refer to this enum:

https://github.com/KratosMultiphysics/Kratos/blob/fe5c313a944845821a3e202c64ed47ee41ddb646/kratos/geometries/geometry_data.h#L76

And the arrays of integration points size then I would go for having more methods added to the enum and making the geometries have only the implemented ones filled and the rest of them are empty with a debug check.

**Comment by loumalouomega** [comment]
> If you refer to this enum:
> 
> https://github.com/KratosMultiphysics/Kratos/blob/fe5c313a944845821a3e202c64ed47ee41ddb646/kratos/geometries/geometry_data.h#L76
> 
> And the arrays of integration points size then I would go for having more methods added to the enum and making the geometries have only the implemented ones filled and the rest of them are empty with a debug check.

I agree in such change
